stylua-vscode/src/download.ts

```diff
@@ -0,0 +1,154 @@
+import * as vscode from "vscode";
+import * as unzip from "unzipper";
+import * as util from "./util";
+import fetch from "node-fetch";
+import { createWriteStream } from "fs";
+import { executeStylua } from "./stylua";
+import { GitHub, GitHubRelease } from "./github";
+
+export class StyluaDownloader {
+  constructor(
+    private readonly storageDirectory: vscode.Uri,
+    private readonly github: GitHub
+  ) {}
+
+  public async ensureStyluaExists(): Promise<string | undefined> {
+    const path = await this.getStyluaPath();
+
+    if (path === undefined) {
+      await vscode.workspace.fs.createDirectory(this.storageDirectory);
+      await this.downloadStyLuaVisual();
+      return await this.getStyluaPath();
+    } else {
+      if (!(await util.fileExists(path))) {
+        vscode.window.showErrorMessage(
+          `The path given for StyLua (${path}) does not exist`
+        );
+        return;
+      }
+
+      try {
+        const currentVersion = (
+          await executeStylua(path, ["--version"])
+        )?.trim();
+        const desiredVersion = util.getDesiredVersion();
+        const release = await this.github.getRelease(desiredVersion);
+        if (
+          currentVersion !==
+          `stylua ${
+            release.tagName.startsWith("v")
+              ? release.tagName.substr(1)
+              : release.tagName
+          }`
+        ) {
+          this.openUpdatePrompt(release);
+        }
+      } catch (err) {
+        vscode.window.showWarningMessage(
+          `Error checking the selected StyLua version, falling back to the currently installed version:\n${err}`
+        );
+        if (!this.github.authenticated) {
+          const option = await vscode.window.showInformationMessage(
+            "Authenticating with GitHub can fix rate limits.",
+            "Authenticate with GitHub"
+          );
+          switch (option) {
+            case "Authenticate with GitHub":
+              if (await this.github.authenticate()) {
+                return this.ensureStyluaExists();
+              }
+          }
+        }
+      }
+
+      return path;
+    }
+  }
+
+  public downloadStyLuaVisual(): Thenable<void> {
+    return vscode.window.withProgress(
+      {
+        cancellable: false,
+        location: vscode.ProgressLocation.Notification,
+        title: "Downloading StyLua",
+      },
+      () => this.downloadStylua()
+    );
+  }
+
+  private async downloadStylua(): Promise<void> {
+    const version = util.getDesiredVersion();
+    const release = await this.github.getRelease(version);
+    const assetFilename = util.getAssetFilenamePattern();
+    const outputFilename = util.getDownloadOutputFilename();
+
+    for (const asset of release.assets) {
+      if (assetFilename.test(asset.name)) {
+        const file = createWriteStream(
+          vscode.Uri.joinPath(this.storageDirectory, outputFilename).fsPath,
```

Comment:
Rather than overwriting a single binary each time we download, should we cache each binary, and name them `stylua-vX.Y.Z`? This means we only need to download a binary once, and if a user changes versions depending on their project often, we don't keep starting download requests.

Maybe this is out of scope for this PR though

Reply:
Another thing, which is also out of scope for this PR, is probably to write the new extension to a temporary location first (incase someone exits VSCode during a download), then move the fully downloaded binary to the wanted location afterwards, for atomicity

Reply:
I think you're right on all counts. Those are good ideas, but out of scope of this PR I think. Most of this code is simply being moved to encapsulate it.